README.md
# XXX IoT Sensor

A core project template for the analysis of current technologies and 
their application to industry - specifically within Internet of Things.

This code base is not complete, but **no** contributions are required.

## General Information

* Free software: Open Software License ("OSL") v. 3.0
* Documentation: To be added


## Features

* Bult-in Hardware Sensor (Raspberry Pi) details monitored and delivered via MQTT
* Temperature, pressure and humidity and other sensor inputs monitored and delivered via MQTT

## Requirements

This code presumes certain hardware is being used.

### Hardware:
* Raspberry Pi 3B or later
* Raspberry SenseHAT

The SenseHAT may be preplaced with the equivalent hardware, but in this case any driver code,
or similar, will need to be created, or installed from the required suppliers.

### Package Requirements

This project requires the following package(s):

| Package | Purpose | Recommended Version |
|------------------|-----------------------------------|----------------|
| `psutils`  | system utilities | v5.8.0 or later |
| `piview`  | Raspberry Pi Information package | v2.0.3 or later |
| `paho-mqtt`  | Python MQTT package  | v1.5.1 or later |
 

Remaining packages are Python 'built-ins'.

### Package Installation

The requirements above may be installed using:

```shell
pip3 install PACKAGE_NAME
```
or using the PyCharm Project Preferences


## Credits


## Copyright

Copyright Adrian Gould, 2021 onwards. 
Licensed under the Open Software License version 3.0

Please credit the author whenever this code is used in any capacity.


HISTORY.md
# History


## 1.0.0

Initial version of the base code


## 0.0.1 Initial Version


Blank project, containing:

- starter folder structure
- sensor-system.py base code
- sensor-tph.py base code
- [[README.md]]
- [[HISTORY.md]]
- [[LICENSE]]


sensor-tph.py
# Project:    xxx-iot-sensor
# Filename:   sensor-tph.py
# Location:   ./
# Author:     STUDENT NAME <STUDENT TAFE EMAIL ADDRESS>
# Created:    28/05/21
# Purpose:
#     This file provides the following features, methods and associated
#     supporting code:
#
#     TODO: STUDENT TO DESCRIBE THE PURPOSE OF THIS FILE
#
# Requirements:
#     An MQTT Server to act as a 'broker' to accept messages and pass onto
#     subscribers to the message topic(s).
#     Python 3.6 or later
#
# Required Packages:
#     This project requires the following Python Packages to be installed:
#         piview
#         paho-mqtt
#


def main():
    import datetime
    import json
    import time

    import paho.mqtt.client as mqtt
    from piview.Host import Host
    from piview.Network import Network

    # MQTT Server details
    mqtt_server = "127.0.0.1"  # Default to local machine
    mqtt_port = 1883  # Default port used by MQTT
    mqtt_time_alive = 60  # Keep Alive time

    #  Get the Pi name, model, revision, serial number and MAC
    sensor_name = Host.name()
    sensor_serial = Host.serial()
    sensor_mac = Network.mac()
    sensor_model = Host.model()
    sensor_revision = Host.revision()

    mqtt_sensor_name = f"{sensor_name}-{sensor_serial}"

    # Define message topic and period between data updates
    topic = "NMTAFE/IoT"
    time_between_updates = 5.0  # seconds

    # set flag for connection status of sensor to MQTT server
    mqtt.Client.connected_flag = False

    def on_connect(client, userdata, flags, rc):
        if rc == 0:
            print("connected OK Returned code=", rc)
            client.connected_flag = True
            # TODO: Use PiView to obtain the sensor data, save to
            #       a variable and then make sure the data is inserted into the
            #       dictionary as shown for the sensor_ip
            sensor_ip = Network.ip()
            # TODO: Add a variable (as defined below) and retrieve the
            #  required details for the Pi using the piview package for
            #  each of the following:
            #       boot_time
            #       sensor_ram = Storage.ram()
            #           ram_total
            #           ram_free
            #           storage
            #           storage_total
            #           storage_free
            #       hw_i2c = Hardware.i2c()
            #           hw_bt
            #           hw_camera
            #           hw_spi

            # Create the data payload to send as the message to the MQTT server
            # TODO: Remember to replace the 'UNKNOWN', 'False' etc with the
            #       correct newly created variable as given above
            data = json.dumps({
                'sensor': mqtt_sensor_name,
                'message': 'connected',
                'model': 'UNKNOWN',
                'ip': sensor_ip,
                'mac': sensor_mac,
                'boot-time': '0000-00-00 00:00:00.0',
                'ram': 0,
                'storage': 0,
                'hw-i2c': False,
                'hw-bt': False,
                'hw-camera': False,
                'hw-spi': False,
                'time': str(datetime.datetime.now())
            })
            client.publish(topic, data)
        else:
            # TODO: Add error code detail display (error codes 1-5)
            print("Bad connection Returned code=", rc)

    # Create MQTT Client connection and callback hooks
    client = mqtt.Client(mqtt_sensor_name, clean_session=True)
    client.on_connect = on_connect
    client.loop_start()

    # Connect to the MQTT Server
    client.connect(mqtt_server, mqtt_port, mqtt_time_alive)

    print('Waiting for connection to MQTT Server...')
    while not client.connected_flag:  # wait in loop
        time.sleep(1)

    # a short delay before getting first readings
    time.sleep(1)

    while True:
        # TODO: Create the code to give readings for each of the SenseHat
        #       sensors listed below. Replace the strings as needed.
        sensehat_temperature = 'SenseHat Temperature Reading'
        sensehat_pressure = 'SenseHat Pressure Reading'
        sensehat_humidity = 'SenseHat Humidity Reading'
        sensehat_accelerometer = 'SenseHat Accelerometer Reading'
        sensehat_compass = 'SenseHat Compass Reading'

        # Create the JSON string to send to the MQTT server
        data = json.dumps({
            'sensor': mqtt_sensor_name,
            'temperature': sensehat_temperature,
            # TODO: pressure reading key-value pair
            # TODO: humidity reading key-value pair
            # TODO: accelerometer reading key-value pair
            # TODO: compass reading key-value pair
            'time': str(datetime.datetime.now()),
        })

        # Send the topic and data to MQTT Server
        client.publish(topic, data)

        # Display data just delivered
        print(f"Published {topic} : {data}")

        # Sleep until next reading required
        time.sleep(time_between_updates)


if __name__ == '__main__':
    main()


sensor-system.py
'''
Project:    xxx-iot-sensor
Filename:   sensor-system.py
Location:   ./
Author:     STUDENT NAME <STUDENT TAFE EMAIL ADDRESS>
Created:    28/05/21
Purpose:
    This file provides the following features, methods and associated
    supporting code:

    TODO: STUDENT TO DESCRIBE THE PURPOSE OF THIS FILE

Requirements:
    An MQTT Server to act as a 'broker' to accept messages and pass onto
    subscribers to the message topic(s).
    Python 3.6 or later

Required Packages:
    This project requires the following Python Packages to be installed:
        piview
        paho-mqtt
            
'''

import datetime
import json
import time

import paho.mqtt.client as mqtt
from piview.CPU import CPU
from piview.Host import Host
from piview.Network import Network

# TODO: Add the required imports from piview to get the GPU temperature,
#       and other details as needed such as Storage, RAM etc

mqtt_server = "127.0.0.1"
mqtt_port = 1883
mqtt_time_alive = 60
# Define the MQTT topic as given below
topic = "NMTAFE/IoT"

#  Get the Pi name, model, revision, serial number, IP and MAC
sensor_name = Host.name()
sensor_serial = Host.serial()
sensor_mac = Network.mac()
sensor_model = Host.model()
sensor_revision = Host.revision()
sensor_ip = Network.ip()
# create the sensor name for MQTT
mqtt_sensor_name = f"{sensor_name}-{sensor_serial}"

# TODO: Once this module is working correctly, change update frequency to 60.0s
update_frequency = 10.0

mqtt.Client.connected_flag = False


def on_connect(client, userdata, flags, rc):
    '''On connection, give details of the sensor including:
        Name, Serial Number, IP Address, Total RAM, Total Storage, Boot Time
        Hardware statuses (i2c, camera, etc) plus time of connection

    :param client:
    :param userdata:
    :param flags:
    :param rc:
    :return:
    '''
    if rc == 0:
        # Connection successful
        print("connected OK Returned code=", rc)
        client.connected_flag = True
        # TODO: Obtain Sensor's Total RAM from piview.Storage
        sensor_ram = 0
        # TODO: Obtain Sensor's Total Storage from piview.Storage
        sensor_storage = 0
        # TODO: Obtain Sensor's I2C Hardware status from piview.Hardware
        sensor_i2c = False
        # TODO: Obtain Sensor's Bluetooth Hardware status from piview.Hardware
        sensor_bt = False
        # TODO: Obtain Sensor's Camera Hardware status from piview.Hardware
        sensor_camera = False
        # TODO: Obtain Sensor's SPI Hardware status from piview.Hardware
        sensor_spi = False
        # TODO: Obtain Sensor's Boot Time from piview.Host
        sensor_boot_time = '0000-00-00 00:00:00.0'

        # Create the payload data to be sent in the message to the MQTT server
        # TODO: Update the items without variables after a colon (:) (They
        #       have dummy values or strings to provide placeholders.
        #       For example the 'ram' : 0
        #       Should read 'ram' : sensor_ram
        data = json.dumps({
            'sensor': mqtt_sensor_name,
            'message': 'connected',
            'model': 'UNKNOWN',
            'ip': sensor_ip,
            'mac': sensor_mac,
            'boot-time': sensor_boot_time,
            'ram': 0,
            'storage': sensor_storage,
            'hw-i2c': False,
            'hw-bt': False,
            'hw-camera': False,
            'hw-spi': sensor_spi,
            'time': str(datetime.datetime.now()),
        })
        client.publish(topic, data)
    else:
        # TODO: Add error code detail display (error codes 1-5)
        print("Bad connection Returned code=", rc)


client = mqtt.Client(mqtt_sensor_name, clean_session=True)
client.on_connect = on_connect
client.loop_start()

client.connect(mqtt_server, mqtt_port, mqtt_time_alive)

print('Waiting for connection to MQTT Server...')
while not client.connected_flag:  # wait in loop
    time.sleep(1)

time.sleep(1)
while True:
    # Add sensor CPU Max Load
    sensor_max_load = CPU.max_load()
    # TODO: Add sensor_cpu_temperature - reading from CPU.temperature
    cpu_temperature = 'Sensor CPU Temperature',
    # TODO: Add sensor_gpu_temperature - reading from GPU.temperature
    gpu_temperature = 'Sensor GPU Temperature',
    # TODO: Add sensor_free_ram and sensor_total_ram
    sensor_free_ram = 'Sensor Free RAM',
    sensor_total_ram = 'Sensor Total RAM',
    # TODO: Add sensor_free_storage and sensor_total_storage
    sensor_free_storage = 'Sensor Free Storage',
    sensor_total_storage = 'Sensor Total Storage',

    # create JSON data to send in message payload
    data = json.dumps({
        'sensor': mqtt_sensor_name,
        'cpu-temp': 'cpu_temperature',
        'gpu-temperature': 'gpu_temperature',
        'cpu-max-load': sensor_max_load,
        "gpu_temperature": "gpu_temperature",
        "sensor-free-ram": "sensor_free_ram",
        "sensor-total-ram": "sensor_total_ram",
        "sensor-free-storage": "sensor_free_storage",
        "sensor-total-storage": "sensor_total_storage",
        'time': str(datetime.datetime.now()),
    })
    client.publish(topic, data)
    print(f"Published {topic} : {data}")
    time.sleep(update_frequency)
